### DISCUSSION TITLE
About the types of colors used in wal

### DISCUSSION THREAD

**Discussion by EldoDebug** [luisbocanegra/kde-material-you-colors#226] [General]
hi
I am developing a non kde desktop environment to use material you
So one thing I was wondering, could you tell me the name of the color you use in wal?

Example:
background = Surface
foreground = OnSurface
color0 = Surface...

I couldn't read and understand the code so I hope you can help me

**Comment by luisbocanegra** [comment]
Only some colors are taken directly from material you, the rest are taken from either the top ranked colors, primary or secondary tones.

```json
"special": {
        "background": "surface",
        "foreground": "surface + tones_Secondary[90] blended by .98",
        "cursor": "onSurface"
    }
```

The rest come from the top ranked colors (`best`) or, if best are less than 7, by iterating `best`, `tones_primary[tone]` and `tones_tertiary[tone]` where tone (lightness) increases by 8 every iteration. This ensures we always get terminal color scheme even if the wallpaper is a single color.

![Screenshot_20240817_110732](https://github.com/user-attachments/assets/82b7bacc-1fe4-4f22-a9ab-cf38b64db8b2)

But what I run instead most of the time is a `custom_colors_list` for the seven ANSI colors which replaces the "best" colors from the image https://github.com/luisbocanegra/kde-material-you-colors/blob/e3e29041883b10c0ab02bbeacd931216ddb2a035/src/kde_material_you_colors/schemeconfigs.py#L25

![Screenshot_20240817_111901](https://github.com/user-attachments/assets/96275210-0cea-411f-b5e3-634de1bc0c4a)

Code that generates the color scheme starts here, a blending to ensure contrast is applied too
https://github.com/luisbocanegra/kde-material-you-colors/blob/e3e29041883b10c0ab02bbeacd931216ddb2a035/src/kde_material_you_colors/schemeconfigs.py#L63

The resultant colors dictionary is generated here https://github.com/luisbocanegra/kde-material-you-colors/blob/e3e29041883b10c0ab02bbeacd931216ddb2a035/src/kde_material_you_colors/schemeconfigs.py#L580

There are some keys like `backgroundIntense` and colors past `color15` because they are used for Konsole but are not used by pywal

```py
self._wal_dark_scheme = {
            "wallpaper": wallpaper_data,
            "alpha": "100",
            "special": {
                "background": pywal_colors_dark[0],
                "backgroundIntense": blendColors(
                    tones_neutral[8], colors_dark["primary"], 0.0
                ),
                "backgroundFaint": blendColors(
                    tones_neutral[8], colors_dark["primary"], 0.35
                ),
                # Normal, ansi 39
                "foreground": blendColors(
                    pywal_colors_dark[0], tones_secondary[90], 0.98
                ),
                # bold, ansi 39
                "foregroundIntense": tones_secondary[90],
                # faint, ansi 39
                "foregroundFaint": blendColors(
                    pywal_colors_dark[0], tones_secondary[90], 0.88
                ),
                "cursor": colors_dark["onSurface"],
            },
            "colors": {
                "color0": pywal_colors_dark[0],
                "color1": pywal_colors_dark[1],
                "color2": pywal_colors_dark[2],
                "color3": pywal_colors_dark[3],
                "color4": pywal_colors_dark[4],
                "color5": pywal_colors_dark[5],
                "color6": pywal_colors_dark[6],
                "color7": pywal_colors_dark[7],
                "color8": pywal_colors_dark_intense[0],
                "color9": pywal_colors_dark_intense[1],
                "color10": pywal_colors_dark_intense[2],
                "color11": pywal_colors_dark_intense[3],
                "color12": pywal_colors_dark_intense[4],
                "color13": pywal_colors_dark_intense[5],
                "color14": pywal_colors_dark_intense[6],
                "color15": pywal_colors_dark_intense[7],
                "color16": pywal_colors_dark_faint[0],
                "color17": pywal_colors_dark_faint[1],
                "color18": pywal_colors_dark_faint[2],
                "color19": pywal_colors_dark_faint[3],
                "color20": pywal_colors_dark_faint[4],
                "color21": pywal_colors_dark_faint[5],
                "color22": pywal_colors_dark_faint[6],
                "color23": pywal_colors_dark_faint[7],
            },
        }
```

It's a mess but is what worked for me.

**Comment by EldoDebug** [comment]
I'm fully understand!!
It's a very clear and wonderful system
Thank you so much!
